src/random-seed/random-seed.c

```diff
@@ -102,57 +185,131 @@ static int run(int argc, char *argv[]) {
+                else {
+                        r = loop_write(random_fd, &mid, sizeof(mid), false);
+                        if (r < 0)
+                                log_debug_errno(r, "Failed to write machine ID to /dev/urandom, ignoring: %m");
+                }
 
                 k = loop_read(seed_fd, buf, buf_size, false);
                 if (k < 0)
-                        r = log_error_errno(k, "Failed to read seed from " RANDOM_SEED ": %m");
-                else if (k == 0) {
-                        r = 0;
+                        log_error_errno(k, "Failed to read seed from " RANDOM_SEED ": %m");
+                else if (k == 0)
                         log_debug("Seed file " RANDOM_SEED " not yet initialized, proceeding.");
-                } else {
+                else {
+                        CreditEntropy lets_credit;
+
                         (void) lseek(seed_fd, 0, SEEK_SET);
 
-                        r = loop_write(random_fd, buf, (size_t) k, false);
-                        if (r < 0)
-                                log_error_errno(r, "Failed to write seed to /dev/urandom: %m");
-                }
+                        lets_credit = may_credit(seed_fd);
 
-                /* Let's also write the machine ID into the random seed. Why? As an extra protection against "golden
-                 * images" that are put together sloppily, i.e. images which are duplicated on multiple systems but
-                 * where the random seed file is not properly reset. Frequently the machine ID is properly reset on
-                 * those systems however (simply because it's easier to notice, if it isn't due to address clashes and
-                 * so on, while random seed equivalence is generally not noticed easily), hence let's simply write the
-                 * machined ID into the random pool too. */
-                z = sd_id128_get_machine(&mid);
-                if (z < 0)
-                        log_debug_errno(z, "Failed to get machine ID, ignoring: %m");
-                else {
-                        z = loop_write(random_fd, &mid, sizeof(mid), false);
-                        if (z < 0)
-                                log_debug_errno(z, "Failed to write machine ID to /dev/urandom, ignoring: %m");
+                        /* Before we credit or use the entropy, let's make sure to securely drop the
+                         * creditable xattr from the file, so that we never credit the same random seed
+                         * again. Note that further down we'll write a new seed again, and likely mark it as
+                         * credible again, hence this is just paranoia to close the short time window between
+                         * the time we upload the random seed into the kernel and download the new one from
+                         * it. */
+
+                        if (fremovexattr(seed_fd, "user.random-seed-creditable") < 0) {
```

Comment:
my thinking was: on old systems where getrandom() doesn't exist we want to do the random seed stuff as we always did, but not credit entropy just like we always did. Hence I figure the random seed should exist always, and always be used, but only when it is marked as "creditable" we should credit entropy to it. This is also important during upgrades: when we upgrade from an old systemd that didn't use getrandom() at all in systemd-random-seed.service to one that does: we shouldn't blindly credit the entropy written that way, but still use it.

So yeah, there are basically three cases where the random seed might exist but we can't really say whether it's safe to apply credit for event though we should write it to /dev/urandom:

1. upgrade from older systemd
2. kernel on previous boot didn't support getrandom(), hence we don't know if pool was initialized
3. if the system was shut down abrubptly in the short time frame at boot where we just credited at the seed from disk to the entropy pool and where we about to read back the new one. In this case it might still be worth writing the entropy to /dev/urandom, but probably not to credit for it...

An alternative idea I had was to just make this two files: /var/lib/systemd/random-seed and /var/lib/systemd/creditable-random-seed, and then rename/replace only the correct one... but this created so much ambiguity regarding what to do when both exist so that i though what i actually want is a boolean marker on the file, and hence opted for xattrs which delivers exactly that...

Reply:
also, is the xattr setting/checking really that complicated? it's a single API cool for reading the bool, and another one for writing it.  which sounds pretty Ok to me?

Reply:
Yep, sounds reasonable.